crates/amaru-consensus/src/chain_forward/client_protocol.rs

```diff
@@ -0,0 +1,245 @@
+                state.add_op(op);
+                if waiting {
+                    if let Some((op, tip)) = state.next_op().await {
+                        waiting = false;
+                        handler.send(Some((op, tip)));
+                    }
+                }
+            }
+            ActoInput::Message(ChainSyncMsg::ReqNext) => {
+                if let Some((op, tip)) = state.next_op().await {
+                    handler.send(Some((op, tip)));
+                } else {
+                    handler.send(None);
+                    waiting = true;
+                }
+            }
+            ActoInput::NoMoreSenders => return Ok(()),
+            ActoInput::Supervision { result, .. } => {
+                return result.map_err(ClientError::HandlerFailure).and_then(|x| x);
+            }
+        }
+    }
+}
+
+/// This actor handles the ReqNext part of the chainsync protocol.
+/// It will ask the parent for the next operation whenever needed.
+/// The parent may respond with None to indicate the need to await, then it will eventually send the next operation.
+async fn chain_sync_handler(
+    mut cell: ActoCell<Option<(ClientOp, Tip)>, impl ActoRuntime>,
+    mut server: chainsync::Server<HeaderContent>,
+    parent: ActoRef<ChainSyncMsg>,
+) -> Result<(), ClientError> {
+    loop {
+        let Some(req) = server.recv_while_idle().await? else {
+            return Err(ClientError::ClientTerminated);
+        };
+        if !matches!(req, ClientRequest::RequestNext) {
+            return Err(ClientError::LateIntersection);
+        };
+        parent.send(ChainSyncMsg::ReqNext);
+
+        if let ActoInput::Message(op) = cell.recv().await {
+            match op {
+                Some((ClientOp::Forward(header), tip)) => {
+                    server
+                        .send_roll_forward(to_header_content(header), tip)
+                        .await?;
+                }
+                Some((ClientOp::Backward(point), tip)) => {
+                    server.send_roll_backward(point, tip).await?;
+                }
+                None => {
+                    server.send_await_reply().await?;
+                    let ActoInput::Message(Some((op, tip))) = cell.recv().await else {
+                        return Ok(());
+                    };
+                    match op {
+                        ClientOp::Forward(header) => {
+                            server
+                                .send_roll_forward(to_header_content(header), tip)
+                                .await?;
+                        }
+                        ClientOp::Backward(point) => {
+                            server.send_roll_backward(point, tip).await?;
+                        }
+                    }
+                }
+            }
+        } else {
+            // parent terminated
+            return Ok(());
+        }
+    }
+}
+
+fn to_header_content(_header: Header) -> HeaderContent {
+    todo!()
```

Comment:
here I’ll need help, no clue how to construct that header content

Reply:
`HeaderContent` is basically just the CBOR serialiazation of the header
```
pub struct HeaderContent {
    pub variant: u8,
    pub byron_prefix: Option<(u8, u64)>,
    pub cbor: Vec<u8>,
}
```

so something like 
```
let header_content = HeaderContent { 
    variant: 1,
    byron_prefix: None,
   cbor: to_cbor(&header)
} 
```

should work

Reply:
thanks! my question was mostly, which variant do we want to produce?

Reply:
It should be 1 AFAIK